### ISSUE TITLE
[RFC] Stabilize versioning and release compatibility

### ISSUE THREAD

**Issue by evan2645** [spiffe/spire#487]
# Background
The current SPIRE development workflow involves tagging release versions off master. Features and bugfixes are developed in branches and merged into master.

Versioning follows an _extremely_ loose interpretation of semantic versioning. Since the projects is still pre-1.0, semantic versioning makes no guarantees about version behavior, however we still adhere to a "non-compat changes bump the minor, and compat changes bump the patch" policy... mostly.

We are reaching a point where this no longer suffices. We are committing major breaking changes to master, and also committing bugfixes and other enhancements that folks running previous release need. In order to enhance usability and foster trust, it is time that we put some guarantees in place.

# Proposal
We should formalize our versioning practices. To start conversation, I propose:
1. Breaking changes to configuration files, Workload API, and Registration API **MUST** result in a minor bump (e.g. `0.5.0` -> `0.6.0`). The Node API is not considered a public API.
1. All other changes **SHOULD** be released with a patch bump (e.g. `0.5.0` -> `0.5.1`).
1. This policy will be updated when 1.0.0 is reached.

## Workflow
In order to accomplish the above proposed policy, a change in the development workflow will be required.

To start conversation, I propose:
1. Master is always on the next minor. For instance, if `0.5.1` is the latest release, then master is tracking `0.6.0`.
1. Release branches track previous releases. For instance, when `0.6.0` is released, a new branch is created named `v0.6`. Non-breaking changes and enhancements should be committed to the `v0.6` branch.
1. Changes made on release branches should _also_ be sent to master, if it makes sense to do so. Alternatively, we can periodically merge into master, however it is unclear if this will result in a mess. The former seems like a safer bet, though introduces some minor developer overhead.

To cut a release:
1. Push a commit updating CHANGELOG.md
1. Tag the commit with the version number (e.g. `v0.6.0`)
1. If cutting from master, push a new release branch (e.g. `v0.6`)
1. Bump the patch level on the release branch
1. If cutting from master, bump the minor level

# Conclusion
This proposal is a request for comment. It is intended to allow ongoing development of both major features and bugfixes without introducing headaches for our uses. Note that this has already happened a handful of time, particularly with regards to changes in configuration file format. Any suggestions that can reduce the overhead of the proposed workflow while still meeting its goals are very welcome.

Finally, there are already major breaking changes present in master. The next release will have to be a minor bump. When we make that bump, whatever workflow we decide on here will be enacted.

**Comment by azdagron** [comment]
How long do we plan on maintaining a release branch? What kind of changes are we allowing to go into release branches after the release has been cut?

**Comment by evan2645** [comment]
> How long do we plan on maintaining a release branch?

I was thinking N - 1. For instance, if 0.5 is the latest released minor, then 0.5 and 0.4 would be supported. The latter would receive bugfixes and security updates only.

> What kind of changes are we allowing to go into release branches after the release has been cut?

The most recently cut release branch will get bugfixes, security updates, and enhancements which don't break compatibility. It may also receive features (e.g. new builtin plugins). The general goal here is to allow quick win updates to be delivered faster, and not be blocked on larger works that are occurring simultaneously.

Friendly reminder that this is just a proposal :) disagreements are welcome.

**Comment by evan2645** [comment]
0.6 release has been cut, and branch has been created. Closing this RFC as "accepted"
